### PR TITLE
non_exhaustive datetime

### PR DESCRIPTION
**Pull request by Manishearth**
Split out of #1789

Progress on #170, #383

### CODE REVIEW

**Comment by Manishearth** [comment]
cc @sffc @zbraniecki who had some opinions too

File: components/datetime/src/mock/time_zone.rs [hunk 1]
```diff
@@ -34,6 +34,7 @@ use core::str::FromStr;
 ///     .expect("Failed to parse a time zone.");
 /// ```
 #[derive(Debug, Default)]
+#[allow(clippy::exhaustive_structs)] // this type is stable
```
Comment (gregtatum) [inline components/datetime/src/mock/time_zone.rs:37]: Question: Why is this struct stable but the others aren't? I don't think I have enough context here. Could you please explain your decision for what gets marked as exhaustive and what doesn't?

File: components/datetime/src/options/length.rs [hunk 2]
```diff
@@ -98,6 +99,27 @@ impl Default for Bag {
...
+    pub fn empty() -> Self {
+        Self {
+            date: None,
+            time: None,
+            preferences: None,
+        }
+    }
+
+    /// Constructs a Bag given a date and time field (preferences set to None)
+    pub fn from_date_time_style(date: Option<Date>, time: Option<Time>) -> Self {
```
Comment (gregtatum) [inline components/datetime/src/options/length.rs:115]: Thought: This appears to be just a style change, and not technically required by this patch. You can still construct a non-exhaustive struct using the old style of construction. However, it seems like this way encourages code that won't break when new fields are added to structs.

File: components/datetime/src/options/preferences.rs [hunk 3]
```diff
@@ -42,6 +42,7 @@ use serde::{Deserialize, Serialize};
 /// ```
 #[derive(Debug, Clone, PartialEq, Copy)]
 #[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
+#[allow(clippy::exhaustive_structs)] // this type is stable
 pub struct Bag {
```
Comment (gregtatum) [inline components/datetime/src/options/preferences.rs:46]: Question: How do we know this is stable?

File: components/icu/README.md [hunk 4]
```diff
@@ -53,11 +53,7 @@ use icu::datetime::{DateTimeFormat, options::length, mock::parse_gregorian_from_
 
 let provider = icu_testdata::get_provider();
 
-let options = length::Bag {
```
Comment (gregtatum) [inline components/icu/README.md:56]: Question: Is this a required change or just a drive-by style change. If it's required, could you please explain.

**Review by gregtatum — CHANGES_REQUESTED** [review]
I'm marking request changes, since I have some questions about this patch before I do an... ahem... exhaustive review.

File: components/datetime/src/options/preferences.rs (hunk 3 above)
Comment (Manishearth) [reply components/datetime/src/options/preferences.rs:46]: ah, it shouldn't be

File: components/icu/README.md (hunk 4 above)
Comment (Manishearth) [reply components/icu/README.md:56]: non_exhaustive doesn't work with FRU (`Struct { field: field, ..foo() }`) syntax

File: components/datetime/src/mock/time_zone.rs (hunk 1 above)
Comment (Manishearth) [reply components/datetime/src/mock/time_zone.rs:37]: note: when I say "stable" here, I mean "stable post 1.0"

I feel like the mock stuff should probably be private post 1.0, actually.

The decision is mostly guesswork, alongside these rules of thumb:

 - inputs to ICU4X APIs (options, etc) are non_exhaustive
 - outputs are _probably_ exhaustive unless we think we will want to add fields
 - provider stuff is always exhaustive/stable since we _serialize_ stably

File: components/datetime/src/options/length.rs (hunk 2 above)
Comment (Manishearth) [reply components/datetime/src/options/length.rs:115]: You cannot use the old style of construction anymore. See https://github.com/rust-lang/lang-team/issues/143

I've added this method to make it pleasant to construct

File: components/icu/README.md (hunk 4 above)
Comment (gregtatum) [reply components/icu/README.md:56]: It seems to here, unless I'm missing something.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fbf3e21153e6f05569fd4538b6b963cc

File: components/datetime/src/options/preferences.rs (hunk 3 above)
Comment (Manishearth) [reply components/datetime/src/options/preferences.rs:46]: (fixing)

File: components/icu/README.md (hunk 4 above)
Comment (Manishearth) [reply components/icu/README.md:56]: That's in the same crate, non_exhaustive only applies for cross-crate stuff

File: components/datetime/src/options/length.rs (hunk 2 above)
Comment (sffc) [inline components/datetime/src/options/length.rs:115]: Suggestion: make the arguments non-optional here.

If `None` is ever required, it should be its own constructor (or manually configured).

File: components/datetime/src/options/length.rs [hunk 5]
```diff
@@ -98,6 +99,27 @@ impl Default for Bag {
     }
 }
 
+impl Bag {
+    /// Constructs a Bag with all fields set to None
+    ///
+    /// Note that the [`Default`] implementation returns long date and time options
```
Comment (sffc) [inline components/datetime/src/options/length.rs:105]: Thought: This is kind-of odd. We should try hard to keep `Default::default()` consistent with `::new()` or `::empty()`.

File: components/datetime/src/skeleton/helpers.rs [hunk 6]
```diff
@@ -81,6 +81,7 @@ const REQUESTED_SYMBOL_MISSING: u32 = 10000;
 /// there is no attempt to add on missing fields. This enum encodes the variants for the current
 /// search for a best skeleton.
 #[derive(Debug, PartialEq, Clone)]
+#[allow(clippy::exhaustive_enums)] // this type is stable
```
Comment (sffc) [inline components/datetime/src/skeleton/helpers.rs:84]: Thought: I don't know why this is exported. It looks like something that should be internal.

File: components/datetime/tests/datetime.rs [hunk 7]
```diff
@@ -568,11 +568,9 @@ fn constructing_datetime_format_with_time_zone_pattern_symbols_is_err() {
     };
     use icu_locid::locale;
 
-    let options = DateTimeFormatOptions::Length(Bag {
-        // Full has time-zone symbols.
```
Comment (sffc) [inline components/datetime/tests/datetime.rs:572]: Nit: Restore comment

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/datetime/src/options/length.rs (hunk 5 above)
Comment (Manishearth) [reply components/datetime/src/options/length.rs:105]: Yeah I agree. I didn't want to change things in this PR but perhaps we shoudl file a followup?

**Review by gregtatum — CHANGES_REQUESTED** [review]
Marking as request changes, mostly for Shane's comments. I'll take a look again after they have been addressed.

File: components/datetime/src/skeleton/helpers.rs (hunk 6 above)
Comment (Manishearth) [reply components/datetime/src/skeleton/helpers.rs:84]: It's doc(hidden), i stuck the allow() at the top level instead

**Review by sffc — APPROVED** [review]

File: components/datetime/src/options/length.rs [hunk 8]
```diff
@@ -210,7 +207,8 @@ pub enum Date {
 /// ```
 /// use icu::datetime::options::length;
 ///
-/// let bag = length::Bag::from_date_time_style(None, Some(length::Time::Medium));
+/// let mut bag = length::Bag::empty();
+/// bag.time = Some(length::Time::Medium);
```
[the remark this answers is not in the document]
Comment (gregtatum) [reply components/datetime/src/options/length.rs:211]: Suggestion (follow-up): +1 to having the valid combinations.

**Review by gregtatum — APPROVED** [review]
Thanks! This works for me, and thanks for explaining why this is needed for cross-module support.

**Comment by Manishearth** [comment]
last commit is the new one, addressing the comment

File: components/datetime/src/options/length.rs [hunk 9]
```diff
@@ -98,6 +95,45 @@ impl Default for Bag {
...
+    pub fn from_date_time_style(date: Date, time: Time) -> Self {
+        Self {
+            date: Some(date),
+            time: Some(time),
+            preferences: None,
+        }
+    }
+
+    /// Constructs a Bag given a date field (preferences and time set to None)
+    pub fn from_date_style(date: Date) -> Self {
```
Comment (gregtatum) [inline components/datetime/src/options/length.rs:120]: praise: Thanks for adding these.

File: components/datetime/src/options/preferences.rs [hunk 10]
```diff
@@ -50,9 +47,19 @@ pub struct Bag {
...
+        Self {
+            hour_cycle: Some(h),
+        }
+    }
+}
+
 /// A user preference for adjusting how the hour component is displayed.
 #[derive(Debug, Clone, Copy, PartialEq)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[allow(clippy::exhaustive_enums)] // this type is stable
```
Comment (gregtatum) [inline components/datetime/src/options/preferences.rs:62]: praise: Thanks for adding these comments, it made it much easier to review.

**Review by gregtatum — APPROVED** [review]
